File: python/src/variant_provider/typecheck.py
"""Conservative static types: reject impossible operations before any probing.

Nullable branches retain union types. Data-dependent failures remain runtime
errors so definitions can intentionally preserve their provider's behavior.
"""

import re


class Type:
    def __init__(self, kinds=(), fields=None, item=None):
        self.kinds = set(kinds)
        self.fields = fields
        self.item = item


def merge(types):
    types = list(types)
    kinds = set().union(*(t.kinds for t in types))
    records = [t for t in types if t.fields is not None]
    fields = None
    if records:
        keys = set().union(*(t.fields for t in records))
        fields = {k: merge(t.fields[k] for t in records if k in t.fields) for k in keys}
    items = [t.item for t in types if t.item is not None]
    return Type(kinds, fields, merge(items) if items else None)


def validate_types(definition, error_type):
    def fail(message):
        raise error_type(message)

    def expect(t, *kinds):
        if t.kinds and not t.kinds.intersection(kinds):
            fail(f"Type mismatch: expected {kinds}, got {sorted(t.kinds)}")

    def literal(v):
        if v is None:
            return Type(["null"])
        if type(v) is bool:
            return Type(["bool"])
        if type(v) is int:
            return Type(["int"])
        if isinstance(v, str):
            return Type(["str"])
        if isinstance(v, list):
            return Type(["list"], item=merge(literal(x) for x in v))
        if isinstance(v, dict):
            return Type(["record"], fields={k: literal(x) for k, x in v.items()})
        fail("Unsupported literal type")

    def placeholders(template, args):
        if not re.fullmatch(r"(?:[^{}]|\{[A-Za-z_][A-Za-z_0-9]*\})*", template):
            fail("Invalid format template")
        if set(re.findall(r"\{([A-Za-z_][A-Za-z_0-9]*)\}", template)) - set(args):
            fail("Missing format argument")

    memo = {}

    def query(name):
        if name not in memo:
            memo[name] = infer(definition["queries"][name]["expr"], {})
        return memo[name]

    def infer(e, s):
        op = e["op"]

        def ev(v):
            return infer(v, s)

        if op == "literal":
            return literal(e["value"])
        if op == "const":
            return literal(definition["constants"][e["name"]])
        if op == "ref":
            return s[e["name"]]
        if op == "query":
            return query(e["name"])
        if op == "env":
            return Type(["str", "null"])
        if op == "let":
            local = dict(s)
            for b in e["bindings"]:
                local[b["name"]] = infer(b["value"], local)
            return infer(e["body"], local)
        if op == "if":
            ev(e["condition"])
            return merge([ev(e["then"]), ev(e["else"])])
        if op in ("list", "concat"):
            items = [ev(x) for x in e["items"]]
            if op == "concat":
                for t in items:
                    expect(t, "list")
                items = [t.item for t in items if t.item is not None]
            return Type(["list"], item=merge(items))
        if op == "record":
            if any(k.startswith("$") for k in e["fields"]):
                fail("Reserved record field")
            return Type(["record"], fields={k: ev(v) for k, v in e["fields"].items()})
        if op == "get":
            t = ev(e["value"])
            expect(t, "record")
            if t.fields is not None and e["key"] not in t.fields:
                fail("Unknown record field")
            return t.fields[e["key"]] if t.fields else Type()
        if op == "index":
            t = ev(e["value"])
            expect(t, "list")
            expect(ev(e["index"]), "int")
            return t.item or Type()
        if op == "seq":
            types = [ev(x) for x in e["items"]]
            return types[-1] if types else Type(["null"])
        if op == "binary":
            a, b = ev(e["left"]), ev(e["right"])
            kind = e["kind"]
            if kind in ("eq", "ne", "lt", "le", "gt", "ge"):
                return Type(["bool"])
            kinds = ("int", "str", "list") if kind == "add" else ("int",)
            expect(a, *kinds)
            expect(b, *kinds)
            if a.kinds and b.kinds and not a.kinds & b.kinds:
                fail("Incompatible arithmetic types")
            return merge([a, b])
        if op in ("not", "is_null"):
            ev(e["value"])
            return Type(["bool"])
        if op == "range":
            for k in ("start", "stop", "step"):
                expect(ev(e[k]), "int")
            return Type(["list"], item=Type(["int"]))
        if op in ("map", "filter"):
            t = ev(e["items"])
            expect(t, "list")
            body = infer(e["body"], {**s, e["var"]: t.item or Type()})
            return Type(["list"], item=body) if op == "map" else t
        if op in ("sort", "unique", "unpack"):
            t = ev(e["value"])
            expect(t, "list")
            return t
        if op == "flatten":
            t = ev(e["value"])
            expect(t, "list")
            item = t.item or Type()
            expect(item, "list")
            return item if item.kinds else Type(["list"])
        if op == "split":
            expect(ev(e["value"]), "str")
            return Type(["list"], item=Type(["str"]))
        if op == "parse_int":
            expect(ev(e["value"]), "str", "int")
            return Type(["int"])
        if op == "parse_version":
            expect(ev(e["value"]), "str")
            return Type(["version"])
        if op == "version_part":
            expect(ev(e["value"]), "version")
            return Type(["int"])
        if op == "format":
            placeholders(e["template"], e["args"])
            for v in e["args"].values():
                expect(ev(v), "str", "int", "bool", "null", "version")
            return Type(["str"])
        if op == "call":
            fn = definition["functions"][e["function"]]
            inputs = iter(e["args"])
            fields = {}
            for arg in fn["args"]:
                kind = (
                    "handle:" + fn["library"]
                    if arg["type"] == "handle"
                    else "str"
                    if arg["type"] == "buffer"
                    else "int"
                )
                if arg["direction"] == "in":
                    expect(ev(next(inputs)), kind)
                else:
                    fields[arg["name"]] = Type([kind])
            return Type(["record"], fields=fields)
        if op == "try":
            return merge([ev(e["body"]), ev(e["fallback"])])
        if op == "finally":
            t = ev(e["body"])
            ev(e["cleanup"])
            return t
        if op in ("warn", "raise"):
            placeholders(
                definition["diagnostics"][e["diagnostic"]]["message"], e["args"]
            )
            for v in e["args"].values():
                expect(ev(v), "str", "int", "bool", "null", "version")
            return Type(["null"]) if op == "warn" else Type()
        if op == "configs":
            for f in definition["features"]:
                t = query(f[e["mode"]])
                expect(t, "list")
                expect(t.item or Type(), "str")
            return Type(["list"], item=Type(["record"]))
        fail("Unknown operation")

    for name in definition["queries"]:
        query(name)

    def handles(t):
        return (
            any(k.startswith("handle:") for k in t.kinds)
            or any(handles(x) for x in (t.fields or {}).values())
            or (t.item is not None and handles(t.item))
        )

    for name in definition["exports"].values():
        if handles(query(name)):
            fail("Native handles cannot be exported")

File: rust/src/typecheck.rs
use serde_json::Value;
use std::collections::{HashMap, HashSet};

#[derive(Clone, Default)]
struct Ty {
    kinds: HashSet<String>,
    fields: Option<HashMap<String, Ty>>,
    item: Option<Box<Ty>>,
}
impl Ty {
    fn new(k: &str) -> Self {
        Self {
            kinds: HashSet::from([k.into()]),
            ..Self::default()
        }
    }
    fn list(t: Ty) -> Self {
        Self {
            item: Some(Box::new(t)),
            ..Self::new("list")
        }
    }
}
fn merge(types: Vec<Ty>) -> Ty {
    let mut result = Ty::default();
    let mut fields: HashMap<String, Vec<Ty>> = HashMap::new();
    let mut items = vec![];
    for t in types {
        result.kinds.extend(t.kinds);
        if let Some(f) = t.fields {
            for (k, v) in f {
                fields.entry(k).or_default().push(v);
            }
        }
        if let Some(i) = t.item {
            items.push(*i);
        }
    }
    if !fields.is_empty() {
        result.fields = Some(fields.into_iter().map(|(k, v)| (k, merge(v))).collect());
    }
    if !items.is_empty() {
        result.item = Some(Box::new(merge(items)));
    }
    result
}
fn expect(t: &Ty, kinds: &[&str]) -> Result<(), String> {
    if !t.kinds.is_empty() && !kinds.iter().any(|k| t.kinds.contains(*k)) {
        Err(format!(
            "Type mismatch: expected {kinds:?}, got {:?}",
            t.kinds
        ))
    } else {
        Ok(())
    }
}
fn literal(v: &Value) -> Ty {
    match v {
        Value::Null => Ty::new("null"),
        Value::Bool(_) => Ty::new("bool"),
        Value::Number(_) => Ty::new("int"),
        Value::String(_) => Ty::new("str"),
        Value::Array(a) => Ty::list(merge(a.iter().map(literal).collect())),
        Value::Object(o) => Ty {
            fields: Some(o.iter().map(|(k, v)| (k.clone(), literal(v))).collect()),
            ..Ty::new("record")
        },
    }
}
fn placeholders(template: &str, args: &Value) -> Result<(), String> {
    let mut rest = template;
    while let Some(start) = rest.find(['{', '}']) {
        if &rest[start..start + 1] != "{" {
            return Err("Invalid format template".into());
        }
        rest = &rest[start + 1..];
        let end = rest.find('}').ok_or("Invalid format template")?;
        let name = &rest[..end];
        if name.is_empty()
            || !name
                .chars()
                .enumerate()
                .all(|(i, c)| c == '_' || c.is_ascii_alphabetic() || (i > 0 && c.is_ascii_digit()))
        {
            return Err("Invalid format template".into());
        }
        if args.get(name).is_none() {
            return Err("Missing format argument".into());
        }
        rest = &rest[end + 1..];
    }
    Ok(())
}
struct Checker<'a> {
    d: &'a Value,
    memo: HashMap<String, Ty>,
}
impl Checker<'_> {
    fn query(&mut self, name: &str) -> Result<Ty, String> {
        if let Some(t) = self.memo.get(name) {
            return Ok(t.clone());
        }
        let expr = self.d["queries"][name]["expr"].clone();
        let t = self.infer(&expr, &HashMap::new())?;
        self.memo.insert(name.into(), t.clone());
        Ok(t)
    }
    fn infer(&mut self, e: &Value, s: &HashMap<String, Ty>) -> Result<Ty, String> {
        let op = e["op"].as_str().unwrap();
        let n = |k: &str| e[k].as_str().unwrap();
        Ok(match op {
            "literal" => literal(&e["value"]),
            "const" => literal(&self.d["constants"][n("name")]),
            "ref" => s[n("name")].clone(),
            "query" => self.query(n("name"))?,
            "env" => merge(vec![Ty::new("str"), Ty::new("null")]),
            "let" => {
                let mut local = s.clone();
                for b in e["bindings"].as_array().unwrap() {
                    let t = self.infer(&b["value"], &local)?;
                    local.insert(b["name"].as_str().unwrap().into(), t);
                }
                self.infer(&e["body"], &local)?
            }
            "if" => {
                self.infer(&e["condition"], s)?;
                merge(vec![self.infer(&e["then"], s)?, self.infer(&e["else"], s)?])
            }
            "list" | "concat" => {
                let mut items = vec![];
                for v in e["items"].as_array().unwrap() {
                    let t = self.infer(v, s)?;
                    if op == "concat" {
                        expect(&t, &["list"])?;
                        items.push(t.item.map(|i| *i).unwrap_or_default());
                    } else {
                        items.push(t);
                    }
                }
                Ty::list(merge(items))
            }
            "record" => {
                let mut fields = HashMap::new();
                for (k, v) in e["fields"].as_object().unwrap() {
                    if k.starts_with('$') {
                        return Err("Reserved record field".into());
                    }
                    fields.insert(k.clone(), self.infer(v, s)?);
                }
                Ty {
                    fields: Some(fields),
                    ..Ty::new("record")
                }
            }
            "get" => {
                let t = self.infer(&e["value"], s)?;
                expect(&t, &["record"])?;
                match t.fields {
                    Some(f) => f.get(n("key")).cloned().ok_or("Unknown record field")?,
                    None => Ty::default(),
                }
            }
            "index" => {
                let t = self.infer(&e["value"], s)?;
                expect(&t, &["list"])?;
                expect(&self.infer(&e["index"], s)?, &["int"])?;
                t.item.map(|i| *i).unwrap_or_default()
            }
            "seq" => {
                let mut t = Ty::new("null");
                for x in e["items"].as_array().unwrap() {
                    t = self.infer(x, s)?;
                }
                t
            }
            "binary" => {
                let a = self.infer(&e["left"], s)?;
                let b = self.infer(&e["right"], s)?;
                if ["eq", "ne", "lt", "le", "gt", "ge"].contains(&n("kind")) {
                    Ty::new("bool")
                } else {
                    let kinds = if n("kind") == "add" {
                        vec!["int", "str", "list"]
                    } else {
                        vec!["int"]
                    };
                    expect(&a, &kinds)?;
                    expect(&b, &kinds)?;
                    if !a.kinds.is_empty() && !b.kinds.is_empty() && a.kinds.is_disjoint(&b.kinds) {
                        return Err("Incompatible arithmetic types".into());
                    }
                    merge(vec![a, b])
                }
            }
            "not" | "is_null" => {
                self.infer(&e["value"], s)?;
                Ty::new("bool")
            }
            "range" => {
                for k in ["start", "stop", "step"] {
                    expect(&self.infer(&e[k], s)?, &["int"])?;
                }
                Ty::list(Ty::new("int"))
            }
            "map" | "filter" => {
                let t = self.infer(&e["items"], s)?;
                expect(&t, &["list"])?;
                let mut local = s.clone();
                local.insert(
                    n("var").into(),
                    t.item.clone().map(|i| *i).unwrap_or_default(),
                );
                let body = self.infer(&e["body"], &local)?;
                if op == "map" {
                    Ty::list(body)
                } else {
                    t
                }
            }
            "sort" | "unique" | "unpack" => {
                let t = self.infer(&e["value"], s)?;
                expect(&t, &["list"])?;
                t
            }
            "flatten" => {
                let t = self.infer(&e["value"], s)?;
                expect(&t, &["list"])?;
                let item = t.item.map(|i| *i).unwrap_or_default();
                expect(&item, &["list"])?;
                if item.kinds.is_empty() {
                    Ty::new("list")
                } else {
                    item
                }
            }
            "split" => {
                expect(&self.infer(&e["value"], s)?, &["str"])?;
                Ty::list(Ty::new("str"))
            }
            "parse_int" => {
                expect(&self.infer(&e["value"], s)?, &["str", "int"])?;
                Ty::new("int")
            }
            "parse_version" => {
                expect(&self.infer(&e["value"], s)?, &["str"])?;
                Ty::new("version")
            }
            "version_part" => {
                expect(&self.infer(&e["value"], s)?, &["version"])?;
                Ty::new("int")
            }
            "format" => {
                placeholders(n("template"), &e["args"])?;
                for v in e["args"].as_object().unwrap().values() {
                    expect(
                        &self.infer(v, s)?,
                        &["str", "int", "bool", "null", "version"],
                    )?;
                }
                Ty::new("str")
            }
            "call" => {
                let f = self.d["functions"][n("function")].clone();
                let mut inputs = e["args"].as_array().unwrap().iter();
                let mut fields = HashMap::new();
                for arg in f["args"].as_array().unwrap() {
                    let kind = match arg["type"].as_str().unwrap() {
                        "handle" => format!("handle:{}", f["library"].as_str().unwrap()),
                        "buffer" => "str".into(),
                        _ => "int".into(),
                    };
                    if arg["direction"] == "in" {
                        expect(&self.infer(inputs.next().unwrap(), s)?, &[&kind])?;
                    } else {
                        fields.insert(arg["name"].as_str().unwrap().into(), Ty::new(&kind));
                    }
                }
                Ty {
                    fields: Some(fields),
                    ..Ty::new("record")
                }
            }
            "try" => merge(vec![
                self.infer(&e["body"], s)?,
                self.infer(&e["fallback"], s)?,
            ]),
            "finally" => {
                let t = self.infer(&e["body"], s)?;
                self.infer(&e["cleanup"], s)?;
                t
            }
            "warn" | "raise" => {
                placeholders(
                    self.d["diagnostics"][n("diagnostic")]["message"]
                        .as_str()
                        .unwrap(),
                    &e["args"],
                )?;
                for v in e["args"].as_object().unwrap().values() {
                    expect(
                        &self.infer(v, s)?,
                        &["str", "int", "bool", "null", "version"],
                    )?;
                }
                if op == "warn" {
                    Ty::new("null")
                } else {
                    Ty::default()
                }
            }
            "configs" => {
                let features = self.d["features"].as_array().unwrap().clone();
                for f in features {
                    let t = self.query(f[n("mode")].as_str().unwrap())?;
                    expect(&t, &["list"])?;
                    expect(&t.item.map(|i| *i).unwrap_or_default(), &["str"])?;
                }
                Ty::list(Ty::new("record"))
            }
            _ => return Err("Unknown operation".into()),
        })
    }
}
fn handles(t: &Ty) -> bool {
    t.kinds.iter().any(|k| k.starts_with("handle:"))
        || t.fields.as_ref().is_some_and(|f| f.values().any(handles))
        || t.item.as_ref().is_some_and(|i| handles(i))
}
pub fn validate(d: &Value) -> Result<(), String> {
    let mut checker = Checker {
        d,
        memo: HashMap::new(),
    };
    for name in d["queries"].as_object().unwrap().keys() {
        checker.query(name)?;
    }
    for name in d["exports"].as_object().unwrap().values() {
        if handles(&checker.query(name.as_str().unwrap())?) {
            return Err("Native handles cannot be exported".into());
        }
    }
    Ok(())
}
